Add tutorial for creating data and running Avivator (#247)
hubmapconsortium/vitessce-image-viewer: tutorial/README.md

> # Data preparatation

Reviewer: ```suggestion
# Data Preparatation
```

Viv supports a subset of the files generated from the `bioformats2raw` + `raw2ometiff` pipeline, described 
[here](https://www.glencoesoftware.com/blog/2019/12/09/converting-whole-slide-images-to-OME-TIFF.html). 

> This guide will demonstrate how to generate a pyramidal Zarr or OME-TIFF with Bio-Formats that can be viewed with 

Reviewer: ```suggestion
This guide demonstrates how to generate a pyramidal Zarr or OME-TIFF with Bio-Formats that can be viewed with 
```

[Avivator](http://avivator.gehlenborglab.org) via HTTP.

### Getting Started

This tutorial requires Bio-Formats [`bioformats2raw`](https://github.com/glencoesoftware/bioformats2raw) and 
[`raw2ometiff`](https://github.com/glencoesoftware/raw2ometiff) command-line tools. It's easiest to install 

> these tools using `conda`, but binaries also are available for download on the corresponding github repositories.

Reviewer 2: I would expect most people to know about conda, but just in case it may be good to add a link
```suggestion
these tools using [`conda`](https://docs.conda.io/projects/conda/en/latest/user-guide/install/), but binaries also are available for download on the corresponding github repositories.
```

```bash
$ conda create --name bioformats python=3.8
$ conda activate bioformats
$ conda install -c ome bioformats2raw raw2ometiff
```

### Input data

Bio-Formats is an incredibly valuable toolkit and supports reading over 
[150 file formats](https://docs.openmicroscopy.org/bio-formats/6.5.1/supported-formats.html). You can choose 
one of your own images, but for the purposes of this tutorial, we will use a multiplexed, high-resolution 

> [Perkin Elmer](https://downloads.openmicroscopy.org/images/Vectra-QPTIFF/perkinelmer/) image made available 

Reviewer 2: ```suggestion
[Perkin Elmer](https://downloads.openmicroscopy.org/images/Vectra-QPTIFF/perkinelmer/) (1.95 GB) image made available 
```

under [CC-BY 4.0](https://creativecommons.org/licenses/by/4.0/) on OME.

```bash
$ wget https://downloads.openmicroscopy.org/images/Vectra-QPTIFF/perkinelmer/PKI_scans/LuCa-7color_Scan1.qptiff 
```

> After the image has finished downloading, there are two options for creating a Avivator/Viv-compliant image. 

Reviewer: ```suggestion
After the image has finished downloading, there are two options for creating an Avivator/Viv-compliant image. 
```

### Pyramid Generation

#### Option 1: Create a Bio-Formats "raw" Zarr

The first option is to use `bioformats2raw` with `--file_type=zarr`. The default "raw" file type is currently 
[n5](https://github.com/saalfeldlab/n5), so the flag is required to generate the zarr-based output. This command will
create the OME-XML metadata along with a pyramidal Zarr for high-resolution images.

```bash
$ bioformats2raw LuCa-7color_Scan1.qptiff LuCa-7color_Scan1/ --file_type=zarr
```

> `bioformats2raw` created the file directory `LuCa-7color_Scan1/` which contains the "raw" bioformats output. The root directory 

Reviewer: ```suggestion
`bioformats2raw` creates the file directory `LuCa-7color_Scan1/` which contains the "raw" bioformats output. The root directory 
```

contains a `METADATA.ome.xml` file along with a `data.zarr/` directory containing the zarr
output. This output can be viewed directly with [Avivator] by serving the top-level directory (`LuCa-7color_Scan1/`) 
over HTTP ([see below](#viewing-in-avivator)).

#### Option 2: Create an OME-TIFF

The second option is to run the complete Bio-Formats pipeline to generate a valid OME-TIFF. 

```bash

> $ bioformats2raw LuCa-7color_Scan1.qptiff n5_tile_directory/ 

Reviewer 3: ```suggestion
$ bioformats2raw LuCa-7color_Scan1.qptiff n5_tile_directory/ --tile_width 512 --tile_height 512
```

I think we should use 512 tile sizes.  This is what Mapbox does, I think, and it's what we do in the portal. I think it's a nice balance between screen sizes and data loading.

Author: I think this a good suggestion, but I want the tutorial to feel very simple. Explicit tile sizes are potentially an optimization, but we haven't actually tested for comparison. 

The default tilesize is `1024` and this is how the IDR is producing their pyramidal data. I'd hesitate to suggest something else unless we have a really convincing reason. We could add a footnote that other parameters (like tilesize) can be specified for `bioformats2raw`, but if `tile_width` and `tile_height` are different viv won't work. Thoughts, @ngehlenborg?

Ideally I'd like the tutorial to be run `bioformats2raw` and `raw2ometiff` without any flags and it just works, but that's not the case.

Reviewer: How about suggesting image tile sizes with a link to the bioformats2raw docs that explain how to set something different from the default?

Author: I'm just hesitant to make a "suggested" tile size since we don't really know. Larger tiles means fewer requests, but maybe larger unnecessary data transfer. It's hard to know.  

I'd rather say "You can adjust tile sizes if desired by ____", but my understanding is the tile size of `1024` is chosen currently because the number of files per directory is enormous for the base image, making it difficult to work with that many objects on local disk. By changing to 512, you increase the number of files per directory by 4.

Author: I just added a note. Please let me know if more is necessary.

Reviewer: We would suggest what has well worked for us (512, 1024). Otherwise people might choose tile sizes that don't work (well) and if we don't offer any guidance, they will just be frustrated and decide that Viv doesn't work for them.

Author: Added. Thanks.

$ raw2ometiff n5_tile_directory/ LuCa-7color_Scan1.ome.tif --compression=zlib
```

> NOTE: Viv currently uses [`geotiff.js`](https://geotiffjs.github.io/) for accessing data from remote TIFFs 
> over HTTP. There is limited compression support in this library, so the `--compression` flag must be specified with
> either `zlib` or `Uncompressed` for viewing in the browser.

### Viewing in Avivator

Avivator is a purely client-side application and relies on access to data over HTTP. Therefore, viewing a dataset locally 
requires a simple web-server. The easist option is to use [`http-server`](https://github.com/http-party/http-server#readme),
which can be installed via `npm` or `Homebrew` if using a Mac:

#### Install `http-server`

```bash
$ npm install --global http-server
# or 
$ brew install http-server
```

> #### Starting a server

Reviewer: ```suggestion
#### Starting a Server
```

From within this directory, start a local server and open [Avivator] in your browser. 

```bash
$ http-server --cors='*' --port 8000 .
```

This command starts a web-server and makes the content in the current directory readable over HTTP. Once the server is running,

> open [Avivator] and paste `http://localhost:8000/LuCa7_color1/` (zarr) or `http://localhost:8000/LuCa7_color1.ome.tif` (OME-TIFF) 

Reviewer 2: ```suggestion
open [Avivator] and paste `http://localhost:8000/LuCa-7color_Scan1/` (zarr) or `http://localhost:8000/LuCa7_color1.ome.tif` (OME-TIFF) 
```

Reviewer 2: Also, when I press enter in the input box, it works, but the URL disappears which seems unexpected

> open [Avivator] and paste `http://localhost:8000/LuCa7_color1/` (zarr) or `http://localhost:8000/LuCa7_color1.ome.tif` (OME-TIFF) 

Reviewer 3: ```suggestion
open [Avivator] and paste `http://localhost:8000/LuCa7_color1/` (zarr) or `http://localhost:8000/LuCa7_color1_Scan1.ome.tif` (OME-TIFF) 
```

> to view the repective pyramids generated above.

Reviewer: ```suggestion
to view the respective pyramids generated above.
```

> open [Avivator] and paste `http://localhost:8000/LuCa-7color_Scan1/` (zarr) or `http://localhost:8000/LuCa7_color1.ome.tif` (OME-TIFF) 

Reviewer 2: ```suggestion
open [Avivator] and paste `http://localhost:8000/LuCa-7color_Scan1/` (zarr) or `http://localhost:8000/LuCa-7color_Scan1.ome.tif` (OME-TIFF) 
```

> open [Avivator] and paste `http://localhost:8000/LuCa-7color_Scan1/` (zarr) or `http://localhost:8000/LuCa-7color_Scan1.ome.tif` (OME-TIFF) 

Reviewer 2: ```suggestion
open [Avivator] and paste [`http://localhost:8000/LuCa-7color_Scan1/`](http://avivator.gehlenborglab.org/?image_url=http://localhost:8000/LuCa-7color_Scan1/) (zarr) or [`http://localhost:8000/LuCa-7color_Scan1.ome.tif`](http://avivator.gehlenborglab.org/?image_url=http://localhost:8000/LuCa-7color_Scan1.ome.tif) (OME-TIFF) 
```

You could even add links with the `image_url` parameter pre-filled for convenience

Reviewer 2: Although this may make copying and pasting more difficult...

> Troubleshooting: Viv relies on cross-origin requests to retrieve data from servers. The `--cors='*'` flag is important to ensure
> that the appropriate `Access-Control-Allow-Origin` response is sent from your local server.

> ### Final note on file formats and OME-Zarr

Reviewer: ```suggestion
### Final Note on File Formats and OME-Zarr
```

The Glencoe software and OME teams hava been clear that the "raw" n5/zarr formats produced by `bioformats2raw` should be considered 
experimental for the time being as intermediates for generating valid OME-TIFFs. Therefore `Option 1` is not as stable as `Option 2` 
for generating images for Avivator/Viv.

However, there is activate community development for a next generation file format (NGFF) called 
[OME-Zarr](https://github.com/ome/omero-ms-zarr/blob/master/spec.md), which can be produced in part by 
running `bioformats2raw --file_type=zarr --dimension-order='XYZCT'`. This will generate a valid multiscale zarr 
which is compatible with OME-Zarr, but is missing some metadata within the zarr hierarchy. 

Aviviator can view the "raw" output as described above, and the *same* multiscale pyramid can also be viewed 
in desktop analysis tools like [`napari`](https://github.com/napari/napari). 

> ### Other examples:

Reviewer: ```suggestion
### Other Examples
```

over HTTP ([see below](#viewing-in-avivator)).

> NOTE: Alternate tile dimensions can be specified with the `--tile_width` and `--tile_height` options. 

> > In our experience, tile sizes of 512 and 1024 (default) have worked well in Viv. For more information 

Reviewer: ```suggestion
> In our experience, tile sizes of 512x512 and 1024x1024 (default) work well. Viv can only handle square tiles. For more information 
```
